### PR TITLE
Add timeout to the delivery spec

### PR DESCRIPTION
**Pull request by slinkydeveloper**
Signed-off-by: Francesco Guardiani <francescoguard@gmail.com>

Added timeout to the delivery spec. For more details: https://github.com/knative/eventing/pull/5149

### CODE REVIEW

File: specs/eventing/spec.md [hunk 1]
```diff
@@ -277,6 +277,7 @@ retrieved from ref.
 | `deadLetterSink` | [`duckv1.Destination`](#duckv1.Destination) | Optional    | The sink receiving event that could not be sent to a `Destination`.                                                                                               |             |
 | `retry`          | `int`                                       | Optional    | The minimum number of retries the sender should attempt when sending an event before moving it to the dead letter sink (when specified) or discarded (otherwise). |             |
 | `backoffPolicy`  | `string`                                    | Optional    | The retry backoff policy (`linear` or `exponential`).                                                                                                             |             |
+| `timeout`        | `string`                                    | Optional    | The timeout of each single request.                                                                                                                               |             |
```
Comment (n3wscott) [inline specs/eventing/spec.md:280]: Note this is a behaviour change for Brokers (all implementations), Triggers, Channels (all implementations), Subscriptions, Sequence, and Parallel.

**Comment by n3wscott** [comment]
I am not a Eventing WG lead, but I feel like this is a large change too close to when we are trying to ship a v1 eventing. Adding this will break a ton of out of tee implementations (channels and brokers).

**Comment by slinkydeveloper** [comment]
I think this is an important API change, but on the other hand this is trivially implementable and requested by a user. I'm even inclined to say that this is more a bug fix than a new feature, given that setting the timeout is quite an important delivery feature, more than the retries. 

I don't agree with the idea that we should stop improving our APIs now, even more because we're going v1 and we want to make sure we support a broad range of use cases and the UX is nice. 

This won't break implementations at all, it will just add a field that some implementation won't support **temporarily**. There are some fields that right now even our reference implementation doesn't fully support/implement , eg `Broker.Delivery`.

**Comment by matzew** [comment]
I'd say, adding this is IMO very important _before_ we ship a `v1` API - this is IMO very important to be added, to avoid potential problems

**Comment by rhuss** [comment]
I think both lines of reasonings (API convergence vs. importance to have individual timeouts) are definitely valid, so it's really a tradeoff. My take is because it is a perfect backwards-compatible update it feels like it is safe to add still to V1. 

The following potential issues should be carefully considered IMO:

* *Bikeshedding issues*: Is the naming `timeout` controversial? --> `timeout` seems to be appropriate when the context is clear, *which* timeout is meant. Should it be `deliveryTimeout`, in case there could be another timeout being added in the future for that specific part of the schema? I don't think this is likely, so I think the name is _safe_
* *Implementation stability*: Is the change critical to the stability of the product? As @n3wscott points out it affects a lot of other resources. Are we totally clear *which* implementation of this duck type are affected and need to be changed and is there someone who takes responsibility? Is the implementation easy, straightforward and timely (including E2E tests) ? Do we need a full implementation everywhere or could it be an option to mark the field optional?
* *Opportunity costs*: If we decide to implement that change, does this cost so much effort that something other, possibly with a higher priority can't be tackled? (considering we are striving for a time-boxed approach for v1 which I'm +1)
* *Creating a precedent*: If we add this change in this phase, could it be a precedent for other change requests to be added as well, and we can't easily draw a line which to add and which not? Are there any of this kind of change that has been already rejected?
* *Future proof*: Does the non-inclusion of this feature would require a quick update of V1 afterwards because this is a highly demanded feature? Are we in danger to dilute the API already? If this is a highly requested feature and we nail down V1 for some time to come, then the `timeout` likely will creep into Eventing via a backdoor (label/annotation) if not added as an optional field.

I think the naming is future-proof, but I can't really assess the other concerns. But my feeling is, adding it as an *optional* field would be a good compromise.

**Comment by slinkydeveloper** [comment]
> Are we totally clear which implementation of this duck type are affected and need to be changed and is there someone who takes responsibility? Is the implementation easy, straightforward and timely (including E2E tests) ? Do we need a full implementation everywhere or could it be an option to mark the field optional?

Delivery spec is a mandatory part of the spec to support for every broker and channel implementation. So all these components already support retries, backoff delay etc, they all just set a default timeout.
For all the components in knative and knative-sandbox org, implementing this field should be quite trivial, because most of them use `eventing/pkg/kncloudevents` to deliver events, so we just need to touch that code in order to support this new feature. For components that doesn't use that package, it should still be trivial to implement, given they already support the other parts of the delivery spec. 

> If we decide to implement that change, does this cost so much effort that something other, possibly with a higher priority can't be tackled?

I don't think so?

> Creating a precedent

I don't understand what do you mean by this. This field came out from a user request, and I care about making users happy, in particular when they have a very valid request like this one, no matter if we're releasing 1.0 or 0.0.1, so when you say "creating a precedent" I feel like you're looking at it from the wrong POV.
If a user comes out with a valid feature request the day before 1.0, I'll push to implement it, no matter the moment. The fact that there is gatekeeping from adding _user requested_ features in specific moments in time, because of preparing for a release and whatnot, is a project important deficiency that has to be addressed ASAP, both process wise and technological wise.

> But my feeling is, adding it as an optional field would be a good compromise.

If your meaning of _optional_ is "an implementation can decide to not support it", then this field is not optional. Implementations that doesn't support it will just ignore it for the moment. But all implementations must support DeliverySpec, hence they'll have to support this field. In other words, it won't break, but it must be implemented at some point.

In general, we don't have a way in eventing to say "this field is optional for an implementer to support".

**Comment by rhuss** [comment]
> so when you say "creating a precedent" I feel like you're looking at it from the wrong POV.

I mean, when there has been already some (user-driven) feature requests like this waiting to be implemented before 1.0 and has been rejected because of 1.0 convergence concerns, then adding the timeout would be a precedent. Not sure whether this is the case though (e.q. if such features exist that has been rejected).

> If a user comes out with a valid feature request the day before 1.0, I'll push to implement it, no matter the moment. 

I don't agree here. For a feature to be added in rush before such an important milestone, it must be carefully considered how it impacts that milestone. There needs to be a time span between 'api change freeze' and hardening / releasing the API IMO. If the change request comes after this freeze (which might have happened implicitly, according to the discussion that I'm aware of), then the feature request must not only be _valid_ but also of highest priority to break that feature-silent time. My question is, whether a delivery timeout qualifies for this urgency (and whether we are already in this hard convergence phase).

**Comment by slinkydeveloper** [comment]
> There needs to be a time span between 'api change freeze' and hardening / releasing the API IMO.

There is no such agreement or public discussion about an eventual api change freeze period.

File: specs/eventing/spec.md (hunk 1 above)
Comment (duglin) [reply specs/eventing/spec.md:280]: Do we need to say what happens if we hit the timeout? e.g. what should be the error seen?

Comment (duglin) [reply specs/eventing/spec.md:280]: Can we get an RFC keyword here? It’s not clear if this is required to be supported or optional.

Comment (duglin) [reply specs/eventing/spec.md:280]: Is this a timeout per attempt (eg like an http timeout) or the total time you hold on to the event including all retries?

**Comment by duglin** [comment]
@slinkydeveloper can you please rebase this?

**Comment by duglin** [comment]
@lionelvillard what are you thoughts on the idea of this feature in general?

**Comment by lionelvillard** [comment]
IMHO this is an important feature, easy to implement (most of it is already implemented, if not everything). so +1 to including it

**Comment by duglin** [comment]
is someone going to drive/own this since slinky has moved on?

**Comment by lionelvillard** [comment]
I can drive this.

/assign

**Comment by duglin** [comment]
@lionelvillard rebase needed

**Comment by lionelvillard** [comment]
closing in favor of https://github.com/knative/specs/pull/76 (to avoid CLA problems)

**Pull request closed without merging**